### ISSUE TITLE
Can this work on Deepnote?

### ISSUE THREAD

**Issue by nathancarter** [gap-system/try-gap-in-jupyter#5]
[Deepnote](http://www.deepnote.com) is a cloud Jupyter provider that's much more powerful than Binder and gives free accounts for everyone, plus extra support for educational clients.  I've used it a lot in data science classes.  They permit you to [bring your own Docker container (scroll down that page)](https://docs.deepnote.com/environment/custom-environments) to run custom Jupyter kernels.  If you do, they have custom [launch buttons](https://www.notion.so/Deepnote-Launch-Buttons-63c642a5e875463495ed2341e83a4b2a) just like Binder does.  It would be great to have this more sophisticated alternative available, if it can be made to work.  Features include persistent projects with easy upload/download of all file types, terminal access, PDF export, real-time collaboration with multiple cursors, comments, free notebook publishing, great UI, etc.

**Comment by olexandr-konovalov** [comment]
That may be interesting to investigate, thanks - maybe something for @ZachNewbery if you're interested to try?

**Comment by ZachNewbery** [comment]
Hey @nathancarter, I had the chance to play around a bit with Deepnote making use of the gapsystem/gap-docker container. Right now, it seems that the Docker container is too large (~5.12GB) for Deepnote's image limit of 5GB.

I did also find out rather late in my exploration that Deepnote blacklists `jupyter` as a package (from [this](https://docs.deepnote.com/environment/custom-initialization)):

> Blacklisted packages There are 2 blacklisted packages: jupyter and jedi. These are packages that interfere with the way Deepnote works and their installation would break your environment. In init.ipynb , we scan the requirements files for those packages and skip their installation if they are there.

so I don't think it's even possible even if we were to trim down the image size unfortunately, but we will try it anyway and see if it does happen to work.

**Comment by olexandr-konovalov** [comment]
So, we would have to provide a custom-built container for Deepnote. First try could be based on https://github.com/gap-system/gap-docker/blob/master/Dockerfile, but don't run full `../bin/BuildPackages.sh` - instead, build only packages needed for JupyterKernel.

**Comment by nathancarter** [comment]
Yeah the reason they don't let you install Jupyter or jedi is because they have their own custom versions of those things that make up the Deepnote UI.  So the idea is actually that you shouldn't HAVE to install those things.  (Actually, I suppose they could have Jupyter running as usual but not the notebook/lab component, replacing that with their own server that talks to kernels...not sure how they chose to do it.)

For instance, if you look at the container definition for Julia at https://docs.deepnote.com/environment/custom-environments/running-your-own-kernel#julia-kernel, it just downloads Julia and the necessary packages, then installs the Julia Jupyter kernel, which assumes a Jupyter installation (of some sort) already exists.  I would assume GAP could do the analogous thing (which might even drop the image size below the 5GB mark?).

**Comment by olexandr-konovalov** [comment]
(oops, accidentally closed - now reopened)

**Comment by ZachNewbery** [comment]
After a bit of playing around, I've got what I believe to be a working version of GAP in Deepnote. I've created a repository [here](https://github.com/ZachNewbery/try-gap-in-deepnote) which stores all the necessary instructions & config to open a Deepnote session using GAP!

**Comment by ZachNewbery** [comment]
I do believe that the changes I make to `init.ipynb` could possibly be made redundant, as the only changes it has are symlink commands, so am having a look at getting them inserted somewhere in the Dockerfile

**Comment by nathancarter** [comment]
Great work!

Is there no way to configure a project so that it remembers the correct container in which the project should be run?  That is, can we get to the point where the user can skip the step of opening the environment tab and manually rebuilding the container?

**Comment by olexandr-konovalov** [comment]
very nice @ZachNewbery - I've tried to follow your instructions at https://github.com/ZachNewbery/try-gap-in-deepnote and it all worked. Let's continue further discussions in that repository's issue tracker.

**Comment by ZachNewbery** [comment]
As a quick update -- have been doing some investigation in a [personal repo](https://github.com/ZachNewbery/try-gap-in-deepnote) which has led to the construction of a docker container (hosted on Github's container repo) that can be added as an image to Deepnote for faster access to the GAP kernel [here](https://github.com/gap-system/gap-docker-deepnote). Atm I'm just writing up some documentation for it, but if you'd like you can use the container using the container address `ghcr.io/gap-system/gap-docker-deepnote:main` :))
